### PR TITLE
Auto save editable fields

### PR DESCRIPTION
**Pull request by richrace**
- Added AutoSave component 
- Removed save buttons 
- Success message is shown when autosaved
- Error message for when input is invalid, or we get an error whilst saving

Note: There are some edge case bugs, but we want to merge this large PR and get it into the hands of users rather than wait until we figure out what the cause is and fix it.

### CODE REVIEW

File: cypress/e2e/case-details/handle-triggers-and-exceptions/edit-fields-with-exceptions/edit-asn.cy.ts [hunk 1]
```diff
@@ -84,7 +84,31 @@ describe("ASN", () => {
     cy.get(".moj-badge").should("not.exist")
   })
 
-  it("Should validate ASN correction and save to updated message in the database", () => {
+  it("Should display error message when ASN is failed to auto-save", () => {
```
Comment (richardcane) [inline cypress/e2e/case-details/handle-triggers-and-exceptions/edit-fields-with-exceptions/edit-asn.cy.ts:87]: Wording

File: cypress/e2e/case-details/handle-triggers-and-exceptions/edit-fields-with-exceptions/edit-asn.cy.ts [hunk 2]
```diff
@@ -98,28 +122,18 @@ describe("ASN", () => {
...
-    // Save correction button should be enabled
-    cy.get("#event-name-error").should("not.exist")
-
-    cy.get("button").contains("Save Correction").click()
-    // Save correction button should be disabled to prevent double clicking
-    cy.get("button").contains("Save Correction").should("be.disabled")
 
     // eslint-disable-next-line cypress/no-unnecessary-waiting
-    cy.wait(500)
+    cy.wait(500) // FIXME
```
Comment (richardcane) [inline cypress/e2e/case-details/handle-triggers-and-exceptions/edit-fields-with-exceptions/edit-asn.cy.ts:136]: 🤔

File: cypress/e2e/case-details/handle-triggers-and-exceptions/edit-fields-with-exceptions/edit-asn.cy.ts [hunk 3]
```diff
@@ -236,8 +247,10 @@ describe("ASN", () => {
   it("should display error when invalid ASN is entered", () => {
     loginAndVisit("/bichard/court-cases/0")
 
-    cy.get("#asn").type("asdf")
+    cy.get("#asn").clear()
+    cy.get("#asn").type("1101ZD01448754")
 
+    cy.get(".Defendant-details-table").find(".warning-icon").should("exist")
```
Comment (richardcane) [inline cypress/e2e/case-details/handle-triggers-and-exceptions/edit-fields-with-exceptions/edit-asn.cy.ts:253]: Class name with capital letter? :(

File: cypress/e2e/case-details/handle-triggers-and-exceptions/edit-fields-with-exceptions/edit-next-hearing-location.cy.ts [hunk 4]
```diff
@@ -341,88 +332,65 @@ describe("NextHearingLocation", () => {
...
+    cy.get("ul.moj-sub-navigation__list").contains("Offences").click()
+    cy.get(".govuk-link").contains("Offence with HO100200 - Unrecognised Force or Station Code").click()
+    cy.get("#next-hearing-location").clear()
+    cy.get("#next-hearing-location").type("B43UY00")
 
-      cy.get("ul.moj-sub-navigation__list").contains("Offences").click()
-      cy.get(".govuk-link").contains("Offence with HO100200 - Unrecognised Force or Station Code").click()
-      cy.get("#next-hearing-location").clear()
-      cy.get("#next-hearing-location").type("      ")
+    cy.get("#success-message").should("not.exist")
```
Comment (richardcane) [inline cypress/e2e/case-details/handle-triggers-and-exceptions/edit-fields-with-exceptions/edit-next-hearing-location.cy.ts:349]: Can there ever be two editable fields visible at once?

File: cypress/e2e/case-details/handle-triggers-and-exceptions/edit-fields-with-exceptions/edit-next-hearing-location.cy.ts [hunk 5]
```diff
@@ -341,88 +332,65 @@ describe("NextHearingLocation", () => {
...
+    cy.get(".warning-icon").should("exist")
+    cy.get("#error-message").contains("Autosave has failed, please refresh").should("exist")
```
Comment (richardcane) [inline cypress/e2e/case-details/handle-triggers-and-exceptions/edit-fields-with-exceptions/edit-next-hearing-location.cy.ts:351]: As above

File: cypress/e2e/case-details/handle-triggers-and-exceptions/edit-fields-with-exceptions/edit-next-hearing-location.cy.ts [hunk 6]
```diff
@@ -341,88 +332,65 @@ describe("NextHearingLocation", () => {
...
+    cy.get("ul.moj-sub-navigation__list").contains("Offences").click()
+    cy.get(".govuk-link").contains("Offence with HO100200 - Unrecognised Force or Station Code").click()
+    cy.get("#next-hearing-location").clear()
+    cy.get("#next-hearing-location").type("B43UY00")
+    cy.get("#success-message").contains("Input saved").should("exist")
 
-      cy.get("#next-hearing-location").clear()
-      cy.get("#next-hearing-location").type("B63AD00")
+    // eslint-disable-next-line cypress/no-unnecessary-waiting
+    cy.wait(500) // FIXME
```
Comment (richardcane) [inline cypress/e2e/case-details/handle-triggers-and-exceptions/edit-fields-with-exceptions/edit-next-hearing-location.cy.ts:375]: 🤔

Comment (C-gyorfi) [reply cypress/e2e/case-details/handle-triggers-and-exceptions/edit-fields-with-exceptions/edit-next-hearing-location.cy.ts:375]: Fixed [35b435d](https://github.com/ministryofjustice/bichard7-next-ui/pull/647/commits/35b435d26afacaccc6c1dc75658740559ded6998)

File: cypress/e2e/case-details/handle-triggers-and-exceptions/edit-fields-with-exceptions/edit-asn.cy.ts (hunk 2 above)
Comment (C-gyorfi) [reply cypress/e2e/case-details/handle-triggers-and-exceptions/edit-fields-with-exceptions/edit-asn.cy.ts]: Fixed in [e90b492](https://github.com/ministryofjustice/bichard7-next-ui/pull/647/commits/e90b4927f8a086494dfd8e35743fe73057ccca1a)

File: cypress/e2e/case-details/handle-triggers-and-exceptions/edit-fields-with-exceptions/edit-asn.cy.ts (hunk 3 above)
Comment (C-gyorfi) [reply cypress/e2e/case-details/handle-triggers-and-exceptions/edit-fields-with-exceptions/edit-asn.cy.ts]: [5b6cebd](https://github.com/ministryofjustice/bichard7-next-ui/pull/647/commits/5b6cebd01d630b480f9b5b406c717fc0a5b91a7f)

Comment (richrace) [reply cypress/e2e/case-details/handle-triggers-and-exceptions/edit-fields-with-exceptions/edit-asn.cy.ts]: This is `styled-components` doing 🙃

File: cypress/e2e/case-details/handle-triggers-and-exceptions/edit-fields-with-exceptions/edit-next-hearing-location.cy.ts (hunk 4 above)
Comment (tausifpatel) [reply cypress/e2e/case-details/handle-triggers-and-exceptions/edit-fields-with-exceptions/edit-next-hearing-location.cy.ts]: Added ID on an editable field row to ensure correct block of error is being tested

File: cypress/e2e/case-details/handle-triggers-and-exceptions/edit-fields-with-exceptions/edit-next-hearing-location.cy.ts (hunk 5 above)
Comment (tausifpatel) [reply cypress/e2e/case-details/handle-triggers-and-exceptions/edit-fields-with-exceptions/edit-next-hearing-location.cy.ts]: Same as above

**Review by richardcane — APPROVED** [review]

**Review by richardcane — APPROVED** [review]

**Pull request merged**